File: src/main/java/dev/ipparser/gui/Export.java
package dev.ipparser.gui;

import dev.ipparser.core.IpUtils;
import dev.ipparser.probe.McProbe;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Map;

/**
 * Pure content builder for the results export file (no Swing dependencies),
 * so the wording and formatting can be unit tested.
 */
public final class Export {

    private Export() {
    }

    public static String build(ExportParams p) {
        StringBuilder sb = new StringBuilder();
        sb.append("IP Parser - scan results").append('\n');
        sb.append("Time: ").append(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date())).append('\n');
        sb.append("Mode: ").append(p.mcMode ? "Minecraft (mcprobe)" : "Telnet (TCP)").append('\n');
        sb.append("Syntax (type=").append(p.syntaxType)
                .append(", input=").append(p.syntaxMode.toLowerCase())
                .append(", CIDR=").append(p.useCidr ? "on" : "off")
                .append("): ").append(p.patternDesc).append('\n');
        sb.append("Port: ").append(p.portSpec).append('\n');
        sb.append("Open ports: ").append(p.openResults.size()).append('\n');
        if (p.external) {
            sb.append("Reachable from outside (public IP): ").append(p.externalCount).append('\n');
        }
        sb.append("------------------------------------------").append('\n');
        for (String res : p.openResults) {
            String line = res;
            if (p.mcMode) {
                McProbe.Result probe = p.mcResults.get(res);
                if (probe != null) {
                    String version = probe.version.isEmpty() ? "?" : probe.version;
                    line = res + "  |  version: " + version
                            + (probe.brand.isEmpty() ? "" : "  |  brand: " + probe.brand)
                            + "  |  players: " + probe.online + "/" + probe.max
                            + (probe.motd.isEmpty() ? "" : "  |  " + probe.motd);
                }
            }
            if (p.external) {
                String ip = res.contains(":") ? res.substring(0, res.lastIndexOf(':')) : res;
                line = line + "  —  " + IpUtils.externalStatus(ip, true, false);
            }
            sb.append(line).append('\n');
        }
        return sb.toString();
    }

    /** All inputs needed to render an export. */
    public static final class ExportParams {
        public boolean mcMode;
        public String syntaxType;
        public String syntaxMode;
        public boolean useCidr;
        public String patternDesc;
        public String portSpec;
        public boolean external;
        public long externalCount;
        public List<String> openResults;
        public Map<String, McProbe.Result> mcResults;
    }
}

File: src/export.rs
//! Pure content builder for the results export file (port of Java `Export`),
//! so the wording and formatting can be unit tested.

use std::collections::HashMap;

use crate::ip_utils;
use crate::mc_probe::McResult;

/// All inputs needed to render an export.
pub struct ExportParams {
    pub mc_mode: bool,
    pub syntax_type: String,
    pub syntax_mode: String,
    pub use_cidr: bool,
    pub pattern_desc: String,
    pub port_spec: String,
    pub external: bool,
    pub external_count: u64,
    pub open_results: Vec<String>,
    pub mc_results: HashMap<String, McResult>,
}

/// Builds the export file content.
pub fn build(p: &ExportParams) -> String {
    let mut sb = String::new();
    sb.push_str("IP Parser - scan results\n");
    sb.push_str(&format!("Time: {}\n", crate::timefmt::now_export_ts()));
    sb.push_str(&format!(
        "Mode: {}\n",
        if p.mc_mode { "Minecraft (mcprobe)" } else { "Telnet (TCP)" }
    ));
    sb.push_str(&format!(
        "Syntax (type={}, input={}, CIDR={}): {}\n",
        p.syntax_type,
        p.syntax_mode.to_lowercase(),
        if p.use_cidr { "on" } else { "off" },
        p.pattern_desc
    ));
    sb.push_str(&format!("Port: {}\n", p.port_spec));
    sb.push_str(&format!("Open ports: {}\n", p.open_results.len()));
    if p.external {
        sb.push_str(&format!(
            "Reachable from outside (public IP): {}\n",
            p.external_count
        ));
    }
    sb.push_str("------------------------------------------\n");
    for res in &p.open_results {
        let mut line = res.clone();
        if p.mc_mode {
            if let Some(probe) = p.mc_results.get(res) {
                let version = if probe.version.is_empty() { "?" } else { &probe.version };
                line = format!(
                    "{}  |  version: {}{}  |  players: {}/{}{}",
                    res,
                    version,
                    if probe.brand.is_empty() {
                        String::new()
                    } else {
                        format!("  |  brand: {}", probe.brand)
                    },
                    probe.online,
                    probe.max,
                    if probe.motd.is_empty() {
                        String::new()
                    } else {
                        format!("  |  {}", probe.motd)
                    }
                );
            }
        }
        if p.external {
            let ip = if res.contains(':') {
                let pos = res.rfind(':').unwrap();
                &res[..pos]
            } else {
                res.as_str()
            };
            line = format!("{}  —  {}", line, ip_utils::external_status(ip, true, false));
        }
        sb.push_str(&line);
        sb.push('\n');
    }
    sb
}
